### ISSUE TITLE
Provide support for subscribing to Azure Event Grid (CloudEvents)

### ISSUE THREAD

**Issue by tomkerkhove** [NuGet/NuGetGallery#7392]
Provide support for subscribing to [CloudEvents](https://cloudevents.io/) allowing users to extend the current NuGet experience.

### Describe the solution you'd like
It would be great if NuGet would allow us to opt-in for receiving events on what is going on in NuGet for the packages that we manage, what is going on in our org, the lifecycle of our API keys, etc.

That way we can fully automate how we manage everything and by using CloudEvents there is no lock-in.

### Scenarios
- **API Key is about to expire/has expired** - Currently emails are already sent out, but if we can receive events we can fully automate the renewal of keys
- **New package has been pushed** - Automatically send out information about new versions being available
- **New people added to an organization** - Automatically send welcome & guidance to them
- etc.

#### But why stop there!
If we can even subscribe to events for individual packages that we can depend on we can automate the updating of our dependencies.

This is already supported with tools like Renovate or Dependabot which are using a pull-based model, but this would allow us to use a push-based more building more reactive apps.

Furthermore, external applications can be more responsive to new packages instead of having to poll your APIs to get the latest. A nice example is the [NuGet indexer that Maarten Balliauw has built](https://blog.maartenballiauw.be/post/2019/07/30/indexing-searching-nuget-with-azure-functions-and-search.html) which could fully rely on events instead.

### Why events over webhooks?
Events provide more granular information about what happened rather than more generic webhooks which allows us to more easily process that information.

**Comment by skofman1** [comment]
Thanks for the suggestion @tomkerkhove ! 
How do you envision the event registration process? What will a user need to provide nuget.org to get events?

**Comment by tomkerkhove** [comment]
User would require to create a "event subscription" which defines the event types they are interested in and the HTTP endpoint to receive the events on.

Bonus points if I can create multiple "subscriptions" allowing us to integrate with multiple internal systems.

**Comment by tomkerkhove** [comment]
Thank you for considering it btw.

**Comment by skofman1** [comment]
> Thank you for considering it btw.
It's an interesting suggestion.

Got it, so when an event occurs we will need to make an HTTP request on O(n) user provided endpoints. 

> Events provide more granular information about what happened rather than more generic webhooks which allows us to more easily process that information.

Read a little about CloudEvents. Isn't this like having webhooks but with a better/well-defined schema?

**Comment by tomkerkhove** [comment]
> Got it, so when an event occurs we will need to make an HTTP request on O(n) user provided endpoints.

Yes and no, if you use a service like Azure Event Grid you, being NuGet gallery/service, can publish events. All user subscriptions are then done on that topic and Event Grid will handle all of the Post in and gives you event filtering out of the box.

Here is some more info :
https://docs.microsoft.com/en-us/azure/event-grid/overview

Long story short, NuGet publishes events and creates the subscriptions wjil Azure Event Grid with Event Domains handles all of the rest. 

Make sense? Happy to discuss on a call of that help more. 

> Read a little about CloudEvents. Isn't this like having webhooks but with a better/well-defined schema?

Yes, definitely. Azure Event Grid supports this out of the box as well but that is just a nice to have.

**Comment by skofman1** [comment]
Sounds good! 
We will add this to our backlog. Implementation timeline will depend on interest from our users.

**Comment by tomkerkhove** [comment]
Thank you @skofman1!
